README.md
# Deskew

Note: Skew is measured in degrees. Deskewing is a process whereby skew is removed by rotating an image by the same amount as its skew but in the opposite direction. This results in a horizontally and vertically aligned image where the text runs across the page rather than at an angle.

The return angle is between -45 and 45 degrees to don't arbitrary change the image orientation.

By using the library you can set the argument `angle_pm_90` to `True` to have an angle between -90 and 90 degrees.

## Skew detection and correction in images containing text

<img src="doc/input.jpeg" width=400px />
<h5>Image with skew</h5>
<img src="doc/sample_output.jpeg" height=300px />
<h5>Image after deskew</h5>

## Cli usage

Get the skew angle:

```
deskew input.png
```

Deskew an image:

```
deskew --output output.png input.png
```

# Lib usage

scikit-image:

```python
import numpy as np
from skimage import io
from skimage.color import rgb2gray
from skimage.transform import rotate

from deskew import determine_skew

image = io.imread('input.png')
grayscale = rgb2gray(image)
angle = determine_skew(grayscale)
rotated = rotate(image, angle, resize=True) * 255
io.imsave('output.png', rotated.astype(np.uint8))
```

OpenCV:

```python
import math
from typing import Tuple, Union

import cv2
import numpy as np

from deskew import determine_skew


def rotate(
        image: np.ndarray, angle: float, background: Union[int, Tuple[int, int, int]]
) -> np.ndarray:
    old_width, old_height = image.shape[:2]
    angle_radian = math.radians(angle)
    width = abs(np.sin(angle_radian) * old_height) + abs(np.cos(angle_radian) * old_width)
    height = abs(np.sin(angle_radian) * old_width) + abs(np.cos(angle_radian) * old_height)

    image_center = tuple(np.array(image.shape[1::-1]) / 2)
    rot_mat = cv2.getRotationMatrix2D(image_center, angle, 1.0)
    rot_mat[1, 2] += (width - old_width) / 2
    rot_mat[0, 2] += (height - old_height) / 2
    return cv2.warpAffine(image, rot_mat, (int(round(height)), int(round(width))), borderValue=background)

image = cv2.imread('input.png')
grayscale = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
angle = determine_skew(grayscale)
rotated = rotate(image, angle, (0, 0, 0))
cv2.imwrite('output.png', rotated)
```

# Debug images

If you get wrong skew angle you can generate debug images, that can help you to tune the skewing detection.

If you install deskew with `pip install deskew[debug_images]` you can get some debug images used for
the skew detection with the function `determine_skew_debug_images`.

To start the investigation you should first increase the `num_peaks` (default `20`) and use
the `determine_skew_debug_images` function.

Then you can try to tune the following arguments `num_peaks`, `angle_pm_90`, `min_angle`, `max_angle`,
`min_deviation` and eventually `sigma`.

Inspired by Alyn: https://github.com/kakul/Alyn


CHANGELOG.md
# Changelog

## [1.3.3](https://github.com/sbrunner/deskew/tree/1.3.3) (2022-08-28)

[Full Changelog](https://github.com/sbrunner/deskew/compare/1.3.2...1.3.3)

#### :bug: Bug Fixes

- Fix angles [\#256](https://github.com/sbrunner/deskew/pull/256) ([sbrunner](https://github.com/sbrunner))

#### :wrench: Build, CI, Tests, Code style

- Better image check [\#254](https://github.com/sbrunner/deskew/pull/254) ([sbrunner](https://github.com/sbrunner))
- Update the changelog [\#250](https://github.com/sbrunner/deskew/pull/250) ([github-actions[bot]](https://github.com/apps/github-actions))

## [1.3.2](https://github.com/sbrunner/deskew/tree/1.3.2) (2022-08-26)

[Full Changelog](https://github.com/sbrunner/deskew/compare/1.3.1...1.3.2)

#### :bug: Bug Fixes

- Better line dev image size [\#253](https://github.com/sbrunner/deskew/pull/253) ([sbrunner](https://github.com/sbrunner))

#### :package: Dependencies update

- Update all patch versions \(patch\) [\#251](https://github.com/sbrunner/deskew/pull/251) ([renovate[bot]](https://github.com/apps/renovate))

## [1.3.1](https://github.com/sbrunner/deskew/tree/1.3.1) (2022-08-22)

[Full Changelog](https://github.com/sbrunner/deskew/compare/1.3.0...1.3.1)

#### :wrench: Build, CI, Tests, Code style

- Update the changelog [\#243](https://github.com/sbrunner/deskew/pull/243) ([github-actions[bot]](https://github.com/apps/github-actions))

#### :package: Dependencies update

- Lock file maintenance [\#249](https://github.com/sbrunner/deskew/pull/249) ([renovate[bot]](https://github.com/apps/renovate))

## [1.3.0](https://github.com/sbrunner/deskew/tree/1.3.0) (2022-08-21)

[Full Changelog](https://github.com/sbrunner/deskew/compare/1.2.0...1.3.0)

#### :sparkles: Enhancements

- Add image to understand the deskewing [\#247](https://github.com/sbrunner/deskew/pull/247) ([sbrunner](https://github.com/sbrunner))

#### :hammer: Refactor

- Remove not needed code that increase the code complexity [\#248](https://github.com/sbrunner/deskew/pull/248) ([sbrunner](https://github.com/sbrunner))

#### :wrench: Build, CI, Tests, Code style

- Remove duplicated [\#246](https://github.com/sbrunner/deskew/pull/246) ([sbrunner](https://github.com/sbrunner))
- Disable the Renovate concurrent limit [\#245](https://github.com/sbrunner/deskew/pull/245) ([sbrunner](https://github.com/sbrunner))
- Update the changelog [\#233](https://github.com/sbrunner/deskew/pull/233) ([github-actions[bot]](https://github.com/apps/github-actions))

#### :package: Dependencies update

- Lock file maintenance [\#244](https://github.com/sbrunner/deskew/pull/244) ([renovate[bot]](https://github.com/apps/renovate))

## [1.2.0](https://github.com/sbrunner/deskew/tree/1.2.0) (2022-08-13)

[Full Changelog](https://github.com/sbrunner/deskew/compare/1.1.0...1.2.0)

#### :sparkles: Enhancements

- Be able to provide min and max angle [\#242](https://github.com/sbrunner/deskew/pull/242) ([sbrunner](https://github.com/sbrunner))

#### :wrench: Build, CI, Tests, Code style

- Add spell on the pull request title [\#241](https://github.com/sbrunner/deskew/pull/241) ([sbrunner](https://github.com/sbrunner))
- Fix mypy types [\#238](https://github.com/sbrunner/deskew/pull/238) ([sbrunner](https://github.com/sbrunner))
- Remove hourly limit [\#236](https://github.com/sbrunner/deskew/pull/236) ([sbrunner](https://github.com/sbrunner))
- Update the changelog [\#227](https://github.com/sbrunner/deskew/pull/227) ([github-actions[bot]](https://github.com/apps/github-actions))

#### :package: Dependencies update

- Lock file maintenance [\#240](https://github.com/sbrunner/deskew/pull/240) ([renovate[bot]](https://github.com/apps/renovate))
- Lock file maintenance [\#239](https://github.com/sbrunner/deskew/pull/239) ([renovate[bot]](https://github.com/apps/renovate))
- Lock file maintenance [\#237](https://github.com/sbrunner/deskew/pull/237) ([renovate[bot]](https://github.com/apps/renovate))
- Update dependency coverage to v6.4.2 [\#235](https://github.com/sbrunner/deskew/pull/235) ([renovate[bot]](https://github.com/apps/renovate))
- Lock file maintenance [\#234](https://github.com/sbrunner/deskew/pull/234) ([renovate[bot]](https://github.com/apps/renovate))
- Update actions/setup-python action to v4 [\#221](https://github.com/sbrunner/deskew/pull/221) ([renovate[bot]](https://github.com/apps/renovate))

## [1.1.0](https://github.com/sbrunner/deskew/tree/1.1.0) (2022-07-11)

[Full Changelog](https://github.com/sbrunner/deskew/compare/1.0.26...1.1.0)

#### :sparkles: Enhancements

- Drop Python 3.7 support [\#228](https://github.com/sbrunner/deskew/pull/228) ([sbrunner](https://github.com/sbrunner))

#### :wrench: Build, CI, Tests, Code style

- Better usage of chore [\#225](https://github.com/sbrunner/deskew/pull/225) ([sbrunner](https://github.com/sbrunner))
- Update the changelog [\#219](https://github.com/sbrunner/deskew/pull/219) ([github-actions[bot]](https://github.com/apps/github-actions))
- Update the changelog [\#216](https://github.com/sbrunner/deskew/pull/216) ([github-actions[bot]](https://github.com/apps/github-actions))

#### :package: Dependencies update

- Lock file maintenance [\#232](https://github.com/sbrunner/deskew/pull/232) ([renovate[bot]](https://github.com/apps/renovate))
- Update dependency numpy to v1.23.1 [\#231](https://github.com/sbrunner/deskew/pull/231) ([renovate[bot]](https://github.com/apps/renovate))
- Update dependency poetry to v1.1.14 [\#230](https://github.com/sbrunner/deskew/pull/230) ([renovate[bot]](https://github.com/apps/renovate))
- Lock file maintenance [\#226](https://github.com/sbrunner/deskew/pull/226) ([renovate[bot]](https://github.com/apps/renovate))
- Update dependency numpy to v1.22.0 \[SECURITY\] [\#223](https://github.com/sbrunner/deskew/pull/223) ([renovate[bot]](https://github.com/apps/renovate))
- Lock file maintenance [\#222](https://github.com/sbrunner/deskew/pull/222) ([renovate[bot]](https://github.com/apps/renovate))

tests/test_deskew.py
import os
from typing import TYPE_CHECKING, Any, Tuple

import cv2
import numpy as np
import pytest
from skimage import io
from skimage.metrics import structural_similarity

from deskew import determine_skew, determine_skew_debug_images

if TYPE_CHECKING:
    NpNdarrayInt = np.ndarray[np.uint8, Any]
else:
    NpNdarrayInt = np.ndarray


def check_image(root_folder, image, name, level=1.0):
    assert image is not None, "Image required"
    expected_name = os.path.join(os.path.dirname(__file__), f"{name}.expected.png")
    result_name = os.path.join(root_folder, f"{name}.result.png")
    diff_name = os.path.join(root_folder, f"{name}.diff.png")
    # Set to True to regenerate images
    if False:
        cv2.imwrite(expected_name, image)
        return
    if not os.path.isfile(expected_name):
        cv2.imwrite(result_name, image)
        cv2.imwrite(expected_name, image)
        assert False, "Expected image not found: " + expected_name
    expected = cv2.imread(expected_name)
    assert expected is not None, "Wrong image: " + expected_name
    score, diff = image_diff(expected, image)
    if diff is None:
        cv2.imwrite(result_name, image)
        assert diff is not None, "No diff generated"
    if score < level:
        cv2.imwrite(result_name, image)
        cv2.imwrite(diff_name, diff)
        assert score >= level, f"{result_name} != {expected_name} => {diff_name} ({score} < {level})"


def image_diff(image1: NpNdarrayInt, image2: NpNdarrayInt) -> Tuple[float, NpNdarrayInt]:
    """Do a diff between images."""

    score, diff = structural_similarity(image1, image2, multichannel=True, full=True)
    diff = (255 - diff * 255).astype("uint8")
    return score, diff


@pytest.mark.parametrize(
    "image,expected_angle",
    [
        ("1", pytest.approx(-1.0, abs=0.01)),
        ("2", pytest.approx(-2.0, abs=0.01)),
        ("3", pytest.approx(-6.0, abs=0.01)),
        ("4", pytest.approx(7.0, abs=0.01)),
        ("5", pytest.approx(3.0, abs=0.01)),
        ("6", pytest.approx(-3.0, abs=0.01)),
        ("7", pytest.approx(3.0, abs=0.01)),
        ("8", pytest.approx(15.0, abs=0.01)),
    ],
)
def test_deskew(image, expected_angle):
    root_folder = f"results/{image}"
    if not os.path.exists(root_folder):
        os.makedirs(root_folder)

    image = io.imread(os.path.join(os.path.dirname(__file__), f"deskew-{image}.png"))
    angle = determine_skew(image)
    assert angle == expected_angle


@pytest.mark.parametrize(
    "min_angle,max_angle,angle_pm_90,num_peaks,expected,postfix",
    [
        (None, None, False, 20, -3, ""),
        (None, None, True, 20, -3, "-90"),
        (-10, 10, False, 20, -3, "-min-max"),
        (5, 10, False, 20, None, "-min-max-positive"),
        (-10, -5, False, 20, None, "-min-max-negative"),
        (35, -35, False, 20, None, "-min-max-inverted"),
        (-10, 10, True, 20, -3, "-90-min-max"),
        (5, 10, True, 20, None, "-90-min-max-positive"),
        (-10, -5, True, 20, None, "-90-min-max-negative"),
        (80, -80, True, 20, None, "-90-min-max-inverted"),
        (None, None, False, 200, -3, "-many-peaks"),
        (None, None, True, 200, -3, "-90-many-peaks"),
    ],
)
def test_determine_skew_debug_images(min_angle, max_angle, angle_pm_90, num_peaks, expected, postfix):
    image = io.imread(os.path.join(os.path.dirname(__file__), "deskew-6.png"))
    angle, debug_images = determine_skew_debug_images(
        image, min_angle=min_angle, max_angle=max_angle, angle_pm_90=angle_pm_90, num_peaks=num_peaks
    )
    for name, debug_image in debug_images:
        print(name)
        check_image("results", debug_image, f"debug-images-{name}{postfix}")
    if expected is None:
        assert angle is None
    else:
        assert angle == pytest.approx(expected, abs=0.01)


deskew/cli.py
import argparse
import sys

import numpy as np
from skimage import io
from skimage.color import rgb2gray
from skimage.transform import rotate

from deskew import determine_skew


def main() -> None:
    """Run the command."""
    parser = argparse.ArgumentParser()

    parser.add_argument("-o", "--output", default=None, help="Output file name")
    parser.add_argument("--sigma", default=3.0, help="The use sigma")
    parser.add_argument("--num-peaks", default=20, help="The used number of peaks")
    parser.add_argument(
        "--num-angles", default=180, help="The used number of angle (determine the pressision)"
    )
    parser.add_argument("--background", help="The used background color")
    parser.add_argument(default=None, dest="input", help="Input file name")
    options = parser.parse_args()

    image = io.imread(options.input)
    grayscale = image if len(image) == 2 else rgb2gray(image)
    angle = determine_skew(grayscale, sigma=options.sigma, num_peaks=options.num_peaks)
    if options.output is None:
        print(f"Estimated angle: {angle}")
    else:
        if options.background:
            try:
                background = [int(c) for c in options.background.split(",")]
            except:  # pylint: disable=bare-except
                print("Wrong background color, should be r,g,b")
                sys.exit(1)
            rotated = rotate(image, angle, resize=True, cval=-1) * 255
            pos = np.where(rotated == -255)
            rotated[pos[0], pos[1], :] = background
        else:
            rotated = rotate(image, angle, resize=True) * 255
        io.imsave(options.output, rotated.astype(np.uint8))


if __name__ == "__main__":
    main()


deskew/__init__.py
import subprocess  # nosec
import tempfile
import warnings
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple

import numpy as np
from skimage.color import rgb2gray, rgba2rgb
from skimage.feature import canny
from skimage.transform import hough_line, hough_line_peaks

if TYPE_CHECKING:
    from typing import TypeAlias

    ImageType: TypeAlias = np.ndarray[np.uint8, Any]
    ImageTypeUint64: TypeAlias = np.ndarray[np.uint8, Any]
    ImageTypeFloat64: TypeAlias = np.ndarray[np.float64, Any]
else:
    ImageType = np.ndarray
    ImageTypeUint64 = np.ndarray
    ImageTypeFloat64 = np.ndarray


def determine_skew_dev(
    image: ImageType,
    sigma: float = 3.0,
    num_peaks: int = 20,
    min_angle: Optional[float] = None,  # -np.pi / 2,
    max_angle: Optional[float] = None,  # np.pi / 2,
    min_deviation: float = np.pi / 180,
    angle_pm_90: bool = False,
) -> Tuple[
    Optional[np.float64],
    Tuple[
        Tuple[ImageTypeUint64, List[List[np.float64]], ImageTypeFloat64],
        Tuple[List[Any], List[np.float64], List[np.float64]],
        Tuple[Dict[np.float64, int], Dict[np.float64, int]],
    ],
]:
    """Calculate skew angle."""

    num_angles = round(np.pi / min_deviation)
    imagergb = rgba2rgb(image) if len(image.shape) == 3 and image.shape[2] == 4 else image
    img = rgb2gray(imagergb) if len(imagergb.shape) == 3 else imagergb
    edges = canny(img, sigma=sigma)
    out, angles, distances = hough_line(edges, np.linspace(-np.pi / 2, np.pi / 2, num_angles, endpoint=False))
    hough_line_out = (out, angles, distances)

    hspace, angles_peaks, dists = hough_line_peaks(
        out, angles, distances, num_peaks=num_peaks, threshold=0.05 * np.max(out)
    )
    hough_line_peaks_out = (hspace, angles_peaks, dists)

    if len(angles_peaks) == 0:
        return None, (hough_line_out, hough_line_peaks_out, ({}, {}))

    freqs_original: Dict[np.float64, int] = {}
    for peak in angles_peaks:
        freqs_original.setdefault(peak, 0)
        freqs_original[peak] += 1

    angles_peaks_corrected = [
        (a % np.pi - np.pi / 2) if angle_pm_90 else ((a + np.pi / 4) % (np.pi / 2) - np.pi / 4)
        for a in angles_peaks
    ]
    angles_peaks_filtred = (
        [a for a in angles_peaks_corrected if a >= min_angle]
        if min_angle is not None
        else angles_peaks_corrected
    )
    angles_peaks_filtred = (
        [a for a in angles_peaks_filtred if a <= max_angle] if max_angle is not None else angles_peaks_filtred
    )
    if not angles_peaks_filtred:
        return None, (hough_line_out, hough_line_peaks_out, ({}, {}))

    freqs: Dict[np.float64, int] = {}
    for peak in angles_peaks_filtred:
        freqs.setdefault(peak, 0)
        freqs[peak] += 1

    sorted_keys = sorted(freqs.keys(), key=freqs.get, reverse=True)  # type: ignore
    max_freq = freqs[sorted_keys[0]]

    max_arr = []
    for sorted_key in sorted_keys:
        if freqs[sorted_key] == max_freq:
            max_arr.append(sorted_key)

    angle = None
    for sorted_key in sorted_keys:
        if freqs[sorted_key] == max_freq:
            angle = sorted_key
            break

    return (
        angle,
        (hough_line_out, hough_line_peaks_out, (freqs_original, freqs)),
    )


def determine_skew_debug_images(
    image: ImageType,
    sigma: float = 3.0,
    num_peaks: int = 20,
    angle_pm_90: bool = False,
    min_angle: Optional[float] = None,
    max_angle: Optional[float] = None,
    min_deviation: float = 1.0,
) -> Tuple[Optional[np.float64], List[Tuple[str, ImageType]]]:
    """Calculate skew angle, and return images useful for debugging."""
    import cv2  # type: ignore # pylint: disable=import-outside-toplevel
    import matplotlib.pyplot as plt  # type: ignore # pylint: disable=import-outside-toplevel
    from matplotlib import cm  # pylint: disable=import-outside-toplevel

    min_angle = np.deg2rad(min_angle) if min_angle is not None else None
    max_angle = np.deg2rad(max_angle) if max_angle is not None else None

    skew_angle, data = determine_skew_dev(
        image,
        sigma=sigma,
        num_peaks=num_peaks,
        min_angle=min_angle,
        max_angle=max_angle,
        min_deviation=np.deg2rad(min_deviation),
        angle_pm_90=angle_pm_90,
    )
    hough_line_data, hough_line_peaks_data, all_freqs = data
    freqs_original, freqs = all_freqs

    skew_angle2 = None if skew_angle is None else (skew_angle + np.pi / 2) % (np.pi) - np.pi / 2

    limits = []
    if min_angle is not None and max_angle is not None:
        min_angle_norm = min_angle % (np.pi / (1 if angle_pm_90 else 2))
        max_angle_norm = max_angle % (np.pi / (1 if angle_pm_90 else 2))
        if min_angle_norm < max_angle_norm:
            min_angle_norm += np.pi / (1 if angle_pm_90 else 2)
        for add in [0.0] if angle_pm_90 else [0.0, np.pi / 2]:
            min_angle_limit = (min_angle_norm + add + np.pi / 2) % np.pi - np.pi / 2
            max_angle_limit = (max_angle_norm + add + np.pi / 2) % np.pi - np.pi / 2
            if min_angle_limit < max_angle_limit:
                limits.append((-np.pi / 2, min_angle_limit))
                limits.append((max_angle_limit, np.pi / 2))
            else:
                limits.append((max_angle_limit, min_angle_limit))

    debug_images = []

    # Hough transform
    hspace, theta, distances = hough_line_data
    del theta, distances
    _, axe = plt.subplots(figsize=(15, 5))

    axe.imshow(
        np.log(1 + hspace),
        cmap=cm.gray,
        aspect="auto",
        extent=(-90, 90, 0, hspace.shape[0]),
    )
    axe.set_title("Hough transform")
    axe.set_xlabel("Angles (degrees)")
    axe.set_ylabel("Distance (pixels)")
    if skew_angle is not None:
        axe.axline((np.rad2deg(skew_angle), 0), (np.rad2deg(skew_angle), 10), color="lightgreen")
        if not angle_pm_90:
            if TYPE_CHECKING:
                assert skew_angle2 is not None  # nosec
            axe.axline((np.rad2deg(skew_angle2), 0), (np.rad2deg(skew_angle2), 10), color="lightgreen")

    for limit_min, limit_max in limits:
        if limit_min != -np.pi / 2:
            axe.axline((np.rad2deg(limit_min), 0), (np.rad2deg(limit_min), 10))
        if limit_max != np.pi / 2:
            axe.axline((np.rad2deg(limit_max), 0), (np.rad2deg(limit_max), 10))

        axe.annotate(
            "ignored",
            xy=(np.rad2deg(limit_min), hspace.shape[0] * 3 / 4),
            xycoords="data",
            xytext=(np.rad2deg(limit_min + limit_max) / 2, hspace.shape[0] * 4 / 5),
            textcoords="data",
            ha="center",
            arrowprops={"arrowstyle": "-", "color": "lightblue"},
        )
        axe.annotate(
            "ignored",
            xy=(np.rad2deg(limit_max), hspace.shape[0] * 3 / 4),
            xycoords="data",
            xytext=(np.rad2deg(limit_min + limit_max) / 2, hspace.shape[0] * 4 / 5),
            textcoords="data",
            ha="center",
            arrowprops={"arrowstyle": "-", "color": "lightblue"},
            color="lightblue",
        )

    plt.tight_layout()
    with tempfile.NamedTemporaryFile(suffix=".png") as file:
        plt.savefig(file.name)
        try:
            subprocess.run(["gm", "convert", "-flatten", file.name, file.name], check=True)  # nosec
        except FileNotFoundError:
            print("Install graphicsmagick to don't have transparent background")

        debug_images.append(("hough_transform", cv2.imread(file.name)))

    # Detected lines
    _, axe = plt.subplots(figsize=(image.shape[0] / 100, image.shape[1] / 100))

    axe.imshow(image, cmap=cm.gray)
    axe.set_ylim((image.shape[0], 0))
    axe.set_axis_off()
    axe.set_title("Detected lines")

    for _, angle, dist in zip(*hough_line_peaks_data):
        (coord0x, coord0y) = dist * np.array([np.cos(angle), np.sin(angle)])
        angle2 = (
            (angle % np.pi - np.pi / 2) if angle_pm_90 else ((angle + np.pi / 4) % (np.pi / 2) - np.pi / 4)
        )
        diff = float(abs(angle2 - skew_angle)) if skew_angle is not None else 999.0
        if diff < 0.001:
            axe.axline((coord0x, coord0y), slope=np.tan(angle + np.pi / 2), linewidth=1, color="lightgreen")
        else:
            axe.axline((coord0x, coord0y), slope=np.tan(angle + np.pi / 2), linewidth=1)
            axe.text(
                coord0x,
                coord0y,
                f"{round(np.rad2deg(angle)*1000)/1000}",
                rotation=np.rad2deg(angle - np.pi / 2),
                rotation_mode="anchor",
                transform_rotates_text=True,
            )

    plt.tight_layout()
    with tempfile.NamedTemporaryFile(suffix=".png") as file:
        plt.savefig(file.name)
        try:
            subprocess.run(["gm", "convert", "-flatten", file.name, file.name], check=True)  # nosec
        except FileNotFoundError:
            print("Install graphicsmagick to don't have transparent background")
        image = cv2.imread(file.name)
        debug_images.append(("detected_lines", cv2.imread(file.name)))

    _, axe = plt.subplots(1, 2, figsize=(15, 6), subplot_kw={"polar": True})

    axe[0].set_title("Original detected angles")
    axe[1].set_title("Corrected angles")

    def fill_polar(axe: Any, freqs: Dict[np.float64, int], half: bool = False) -> None:
        axe.scatter(freqs.keys(), freqs.values())
        axe.set_theta_zero_location("N")
        axe.grid(True)
        if half:
            axe.set_thetamin(-45)
            axe.set_thetamax(45)
        else:
            axe.set_thetamin(-90)
            axe.set_thetamax(90)

        if skew_angle is not None:
            axe.axvline(skew_angle, color="lightgreen")
            if not angle_pm_90 or not half:
                axe.axvline(skew_angle2, color="lightgreen")

        for limit_min, limit_max in limits:
            if limit_min != -np.pi / 2 and (not half or -np.pi / 4 < limit_min < np.pi / 4):
                axe.axvline(limit_min)
            if limit_max != np.pi / 2 and (not half or -np.pi / 4 < limit_max < np.pi / 4):
                axe.axvline(limit_max)

    fill_polar(axe[0], freqs_original)
    fill_polar(axe[1], freqs, not angle_pm_90)

    plt.tight_layout()
    with tempfile.NamedTemporaryFile(suffix=".png") as file:
        plt.savefig(file.name)
        try:
            subprocess.run(["gm", "convert", "-flatten", file.name, file.name], check=True)  # nosec
        except FileNotFoundError:
            print("Install graphicsmagick to don't have transparent background")
        image = cv2.imread(file.name)
        debug_images.append(("polar_angles", cv2.imread(file.name)))

    return None if skew_angle is None else np.rad2deg(skew_angle), debug_images


def determine_skew(
    image: ImageType,
    sigma: float = 3.0,
    num_peaks: int = 20,
    num_angles: Optional[int] = None,
    angle_pm_90: bool = False,
    min_angle: Optional[float] = None,
    max_angle: Optional[float] = None,
    min_deviation: float = 1.0,
) -> Optional[np.float64]:
    """
    Calculate skew angle.

    Return None if no skew will be found
    """
    if num_angles is not None:
        min_deviation = 180 / num_angles
        warnings.warn("num_angles is deprecated, please use min_deviation", DeprecationWarning)

    angle, _ = determine_skew_dev(
        image,
        sigma=sigma,
        num_peaks=num_peaks,
        min_angle=np.deg2rad(min_angle) if min_angle is not None else None,
        max_angle=np.deg2rad(max_angle) if max_angle is not None else None,
        min_deviation=np.deg2rad(min_deviation),
        angle_pm_90=angle_pm_90,
    )
    return None if angle is None else np.rad2deg(angle)
